### PR TITLE
[10.0] Fix l10n ch QR-IBAN validation

### PR DESCRIPTION
**Pull request by santostelmo**
Only IBANs starting with CH and LI can be considered QR IBANs

### CODE REVIEW

File: l10n_ch_qr_bill/models/res_bank.py [hunk 1]
```diff
@@ -24,6 +24,18 @@ class ResPartnerBank(models.Model):
         )
     )
 
+    @api.multi
+    @api.constrains('l10n_ch_qr_iban')
```
Comment (carlosserratoro) [inline l10n_ch_qr_bill/models/res_bank.py:27]: You constrain depending on ```l10n_ch_qr_iban``` but you are working in the code only with ```l10n_ch_qrr```. Does ```l10n_ch_qr_iban``` depend on the value set on ```l10n_ch_qrr```?

Comment (santostelmo) [reply l10n_ch_qr_bill/models/res_bank.py:27]: Thanks carlos, this was a mistake. Changing. it should be `l10n_ch_qr_iban` used in the method.

**Review by carlosserratoro — APPROVED** [review]

File: l10n_ch_qr_bill/models/res_bank.py [hunk 2]
```diff
@@ -24,6 +24,18 @@ class ResPartnerBank(models.Model):
...
+    def _check_ch_li_l10n_ch_qr_iban(self):
+        """Check if field is populated with Swiss / Liechtenstein qr iban"""
```
Comment (yvaucher) [inline l10n_ch_qr_bill/models/res_bank.py]:
```suggestion
        """Check if field is populated with Swiss / Liechtenstein QR-IBAN"""
```

File: l10n_ch_qr_bill/models/res_bank.py [hunk 3]
```diff
@@ -24,6 +24,18 @@ class ResPartnerBank(models.Model):
         )
     )
 
+    @api.multi
```
Comment (yvaucher) [inline l10n_ch_qr_bill/models/res_bank.py]: `api.constraint` are `api multi` by default. No need to specify it.

File: l10n_ch_qr_bill/models/res_bank.py [hunk 4]
```diff
@@ -69,6 +81,11 @@ def _is_qr_iban(self):
         """
         return (
             self.acc_type == "iban"
-            and self._check_qr_iban_range(self.sanitized_acc_number)
-            or self.l10n_ch_qr_iban
+            and (
+                self.l10n_ch_qr_iban and
+                self.l10n_ch_qr_iban.startswith(('CH', 'LI'))
+            ) or (
```
Comment (yvaucher) [inline l10n_ch_qr_bill/models/res_bank.py]: This should already be validated by the constraint.

```suggestion
            and self.l10n_ch_qr_iban
            or (

File: l10n_ch_qr_bill/models/res_bank.py [hunk 5]
```diff
@@ -24,6 +24,18 @@ class ResPartnerBank(models.Model):
...
+    @api.constrains('l10n_ch_qr_iban')
+    def _check_ch_li_l10n_ch_qr_iban(self):
+        """Check if field is populated with Swiss / Liechtenstein qr iban"""
+        for record in self:
+            if (
+                record.l10n_ch_qr_iban
+                and not record.l10n_ch_qr_iban.startswith(('CH', 'LI'))
+            ):
+                raise ValidationError(_(
+                    "Not valid Switzerland or Liechtenstein QR IBAN."))
```
Comment (yvaucher) [inline l10n_ch_qr_bill/models/res_bank.py]: Please use `_validate_qr_iban` instead.

And move the CH LI check in there.

It seems `_validate_qr_iban` is there for that, the constraint was missing though.

```suggestion
            if record.l10n_ch_qr_iban:
                self._validate_qr_iban(record.l10n_ch_qr_iban)
```

**Review by yvaucher — CHANGES_REQUESTED** [review]
Would work but we can use `_validate_qr_iban` for the purpose of the constraint.

File: l10n_ch_qr_bill/models/res_bank.py (hunk 5 above)
Comment (santostelmo) [reply l10n_ch_qr_bill/models/res_bank.py]: Done

File: l10n_ch_qr_bill/models/res_bank.py [hunk 6]
```diff
@@ -35,6 +41,9 @@ def _check_qr_iban_range(self, iban):
                 and 30000 <= int(iid) <= 31999)
 
     def _validate_qr_iban(self, qr_iban):
+        if qr_iban and not qr_iban.startswith(('CH', 'LI')):
+            raise ValidationError(_(
+                "Not valid Switzerland or Liechtenstein QR IBAN."))
```
Comment (yvaucher) [inline l10n_ch_qr_bill/models/res_bank.py]:
```suggestion
                "Not a valid Switzerland or Liechtenstein QR-IBAN."))
```

**Comment by santostelmo** [comment]
@yvaucher rebased

**Review by yvaucher — APPROVED** [review]

**Pull request merged**